### PR TITLE
Vfolders namespace

### PR DESCRIPTION
**Pull request by phlax**
addresses concerns raised in #5217 (Namespacing problems with virtualfolders)

### CODE REVIEW

File: pootle/apps/virtualfolder/views.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+
```
Comment (unho) [inline pootle/apps/virtualfolder/views.py]: Nit: only one blank line before imports.

File: tests/views/vf.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    # tp browse views
+    tp_browse_view = reverse(
+        "vfolder-pootle-tp-browse",
+        kwargs=dict(
+            vfolder_name="VF_NAME",
+            language_code="LANG_CODE",
+            project_code="PROJ_CODE"))
+    assert (
+        tp_browse_view
+        == u'/++vfolder/VF_NAME/LANG_CODE/PROJ_CODE/')
```
Comment (unho) [inline tests/views/vf.py]: This fits in a single line

Comment (phlax) [inline tests/views/vf.py]: its clearer this way tho

File: tests/views/vf.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        tp_translate_subdir_store_view
+        == u'/++vfolder/VF_NAME/LANG_CODE/PROJ_CODE/translate/SOME/SUBDIR/FILE.EXT')
+
+
+@pytest.mark.pootle_vfolders
+@pytest.mark.django_db
+def test_views_vf_translate(vfolder_views, settings):
+    from tp import _test_browse_view, _test_translate_view
+    test_type, tp, request, response, kwargs = vfolder_views
+    if test_type == "browse":
```
Comment (phlax) [inline tests/views/vf.py]: this is not necessary

File: pootle/urls.py [hunk 4]
```diff
@@ -13,7 +13,13 @@
 from pootle.core.delegate import url_patterns
 
 
-urlpatterns = [
+urlpatterns = []
+
+# Allow url handlers to be overriden by plugins
+for delegate_urls in url_patterns.gather().values():
```
Comment (unho) [inline pootle/urls.py]: Just curious if this registers the URL handlers in a specific order.

File: pootle/apps/virtualfolder/views.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+from django.utils.functional import cached_property
+
+from pootle_translationproject.views import TPTranslateView, TPTranslateStoreView
```
Comment (unho) [inline pootle/apps/virtualfolder/views.py]: Shouldn't the order for this be inverted? `TPTranslateStoreView, TPTranslateView`

File: pootle/apps/virtualfolder/views.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    @property
+    def path(self):
+        return self.extracted_path[1]
+
+    @property
+    def vfolder_pk(self):
+        vfolder = self.extracted_path[0]
+        if vfolder:
+            return vfolder.pk
+        return ""
```
Comment (unho) [inline pootle/apps/virtualfolder/views.py]: `""` ? Why not `None`?

Comment (unho) [inline pootle/apps/virtualfolder/views.py]: nvm

File: pootle/apps/virtualfolder/urls.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+from .views import get_vfolder_units, VFolderTPTranslateView
+
+
+vfolder_urlpatterns = [
+    # TP Translate
+    url(r'(?P<language_code>[^/]*)/'
+        r'(?P<project_code>[^/]*)/translate/(?P<dir_path>(.*/)*)?/?',
+        VFolderTPTranslateView.as_view(),
+        name='vfolder-pootle-tp-translate'),
```
Comment (unho) [inline pootle/apps/virtualfolder/urls.py]: This doesn't follow the urlpattern names agreements. Use `pootle-virtualfolder-translate` instead.

File: pootle/apps/virtualfolder/urls.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    url(r'xhr/units/$',
+        get_vfolder_units,
+        name='vfolder-pootle-xhr-units')]
```
Comment (unho) [inline pootle/apps/virtualfolder/urls.py]: Same here re naming.

File: pootle/apps/virtualfolder/urls.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+from django.conf.urls import include, url
+
+from pootle_store.urls import unit_xhr_urlpatterns
+
+from .views import get_vfolder_units, VFolderTPTranslateView
+
+
+vfolder_urlpatterns = [
+    # TP Translate
+    url(r'(?P<language_code>[^/]*)/'
```
Comment (unho) [inline pootle/apps/virtualfolder/urls.py]: Why not include here the `++vfolder/vfname`? That way it should be easier to generate this URLs, for example in https://github.com/translate/pootle/pull/5221/commits/79b56314f997cf934106394928baa65b111a59e6#diff-2595359c0c549cb2373dc7ec23dc8cb2R415

File: pootle/apps/pootle_translationproject/views.py [hunk 10]
```diff
@@ -365,48 +365,13 @@ class TPTranslateView(TPDirectoryMixin, TPTranslateBaseView):
...
-        return self.extracted_path[1]
+        return self.request_path
 
     @property
-    def vfolder_pk(self):
-        vfolder = self.extracted_path[0]
-        if vfolder:
-            return vfolder.pk
-        return ""
+    def display_vfolder_priority(self):
```
Comment (unho) [inline pootle/apps/pootle_translationproject/views.py]: This previously was above `path` property. Only by chance noticed that it was moved down. If you want to reposition please do in separate commit.

File: pootle/apps/virtualfolder/urls.py (hunk 9 above)
Comment (phlax) [inline pootle/apps/virtualfolder/urls.py]: its more reusable to put the prefix in the include  - see below

that way the views can be used with other/out prefix etc

File: pootle/apps/pootle_translationproject/views.py (hunk 10 above)
Comment (phlax) [inline pootle/apps/pootle_translationproject/views.py]: im not sure they are needed - ill check

File: pootle/apps/virtualfolder/urls.py (hunk 9 above)
Comment (phlax) [inline pootle/apps/virtualfolder/urls.py]: and less repitition

File: pootle/core/decorators.py [hunk 11]
```diff
@@ -132,15 +131,7 @@ def set_resource(request, path_obj, dir_path, filename):
...
 
-    vfolder = None
     clean_pootle_path = pootle_path
-    if 'virtualfolder' in settings.INSTALLED_APPS:
-        from virtualfolder.helpers import extract_vfolder_from_path
-        # Get a clean pootle path for retrieving the directory or store.
-        # A clean pootle path is a pootle path without any virtual folder name on
-        # it. For example /af/test_vfolders/browser/chrome/ is the corresponding
-        # clean pootle path for /af/test_vfolders/browser/vfolder8/chrome/
-        vfolder, clean_pootle_path = extract_vfolder_from_path(pootle_path)
```
Comment (unho) [inline pootle/core/decorators.py]: Given that this is dropped you can probably simplify the surrounding code a bit more now that `clean_pootle_path` and `pootle_path` are the same.

File: pootle/static/js/browser.js [hunk 12]
```diff
@@ -65,9 +65,15 @@ function navigateTo(languageCode, projectCode, resource) {
...
-
-  const parts = ['', newLanguageCode, projectCode, action, newResource];
-  const urlParts = parts.filter((p, i) => i === 0 || p !== '');
+  let urlParts = [];
+  if (resource.substr(0, 2) === '++') {
+    urlParts = resource.split('/');
+    urlParts.splice(2, 0, newLanguageCode, projectCode, action);
+    urlParts.splice(0, 0, '');
+  } else {
+    const parts = [newLanguageCode, projectCode, action, newResource];
```
Comment (unho) [inline pootle/static/js/browser.js]: Why dropping the  `''`?

Comment (phlax) [inline pootle/static/js/browser.js]: i dont follow?

Comment (unho) [inline pootle/static/js/browser.js]: In the previous code it was:

`const parts = ['', newLanguageCode, projectCode, action, newResource];`

while in the new it is:

`const parts = [newLanguageCode, projectCode, action, newResource];`

Comment (phlax) [inline pootle/static/js/browser.js]: good spot!

File: pootle/core/decorators.py [hunk 13]
```diff
@@ -132,26 +131,16 @@ def set_resource(request, path_obj, dir_path, filename):
...
-
     if filename:
         pootle_path = pootle_path + filename
-        clean_pootle_path = clean_pootle_path + filename
+        pootle_path = pootle_path + filename
```
Comment (unho) [inline pootle/core/decorators.py]: Repeated code

File: pootle/core/decorators.py (hunk 11 above)
Comment (unho) [inline pootle/core/decorators.py]: This is addressed.

**Comment by unho** [comment]
lgtm

**Comment by phlax** [comment]
massive **thanks** for all your hard working reviewing and explaining our bugs back to us @unho !
